go/vt/mysqlctl/schema.go

```diff
@@ -75,11 +85,36 @@ func (mysqld *Mysqld) GetSchema(ctx context.Context, dbName string, tables, excl
 		return sd, nil
 	}
 
-	sd.TableDefinitions = make([]*tabletmanagerdatapb.TableDefinition, 0, len(qr.Rows))
+	filter, err := tmutils.NewTableFilter(tables, excludeTables, includeViews)
+	if err != nil {
+		return nil, err
+	}
+
+	type schemaResult struct {
```

Comment:
@sougou, based on what i think you're saying, i believe you're suggesting something like this sketch:

```go
var tds
var tdsMx

for {
  go func() {
   res := collectSchema()
   tdsMx.Lock()
   tds = append(tds, res)
   tdsMx.Unlock()
  }
}

```

is that right?

while that sketch may happen to work, i don't think the ownership model looks right on it. it seems like no goroutine owns the `tds`, and so it's not really guaranteed to work according to spec.

primitives are ok assuming they're read-only. but something like a structured type doesn't seem to be guaranteed.

hence i wrote this to collect results and serialize them, as there's only one gorouting w/ ownership of `tds`.

Reply:
per discussion w/ @systay and @shlomi-noach, we're going with this gist:

```
func ConcurrentCollect() []interface{} {
	var tds []interface{}
	var tdsMx sync.Mutex
	wg := sync.WaitGroup{}
	for _, chunk := range chunks {
		wg.Add(1)
		go func() {
			defer wg.Done()
			res := collectSchema(chunk)
			tdsMx.Lock()
			tds = append(tds, res)
			tdsMx.Unlock()
		}()
	}
	wg.Wait()
	return tds
}
```

i will followup with changing the code to do this.

Reply:
If order is still important, you can rewrite as follows:
```golang
func ConcurrentCollect() []interface{} {
	var tdsMap map[int]interface{} {}
	var tdsMx sync.Mutex
	wg := sync.WaitGroup{}
	for i, chunk := range chunks {
		wg.Add(1)
		go func() {
			defer wg.Done()
			res := collectSchema(chunk)
			tdsMx.Lock()
			tdsMap[i] = res
			defer tdsMx.Unlock()
		}()
	}
	wg.Wait()

	tds  := make([]interface{}, len(tdsMap))
	for i, td := range tdsMap {
		tds[i] = td
	}
	return tds
}
```

Reply:
Looks like you converged on what I had in mind with @systay. As mentioned before, order is not important.

Reply:
@sougou, playing around with this code some more, i still think sticking w/ the single channel strategy is the right way to do it.

otherwise, i still need to also keep more state outside like an `someErr` variable, and i still need to get the column info from `getPrimaryKeyColumns` separately and then reconstruct the final result.

then i need to handle errors if any come up during the `collectSchema`. so i think my current single channel strategy is the more elegant strategy.

Reply:
My main concern is not the channel per se. It's the extra complexity introduced to maintain order, and that is unnecessary. Let's see if the code reads better once you remove that.

Also, I realize that there are no tests to prove that the new code works. With the instabilities we faced in 6.0, writing of tests is now mandatory.

PS: There are many idiomatic patterns in vitess codebase that use the waitgroup approach. Here's an example that also includes error handling: https://github.com/vitessio/vitess/blob/1f5484748be76058f509ac67c90c7748a0cae383/go/vt/wrangler/traffic_switcher.go#L1030-L1045.

Reply:
@sougou, i can change the behavior to be orderless, but at the moment, ALL the tests in vitess depend on the order being constant. i think changing behavior like that should be a separate PR, and would have to be pretty extensive.

Reply:
i've created a separate version of this PR that has the mutex version previously mentioned. i still think that order changing is a behavior change that deserves it's own PR.

https://github.com/vitessio/vitess/pull/6491